proposals/0002-gep.md

```diff
@@ -0,0 +1,216 @@
+---
+MEP: 0002
+Title: Unifed API
+Discussion: Link
+Implementation: Link
+---
+
+# Unified API
+
+## Abstract
+
+This GEP discusses the unified API that handles the interaction with LLM providers. This represents the feature set for MVP.
+
+## Motivation
+
+
+The Unified API abstracts much of the overhead of interacting with LLM providers.
+It stores all the provider specific logic in a single place and allowing developers to focus on the core logic. It also introduces durability to LLM apps. Additionally, unifying the API allows optimizations to be leveraged such as latency and cost improvements.
+
+
+### Requirements
+
+- R1: Handles all provider specific logic
+- R2: Easily maintained
+- R3: API schemas must unify common LLM request params (e.g. topP, N, etc)
+- R4: API routes must unify common LLM endpoints/API
+- R5: API schemas should allow to customize provider-specific params (e.g. a different messages for OpenAI/Anthropic)
+- R6: Focus on building out a single endpoint for /Chat. Most apps will be using this and then we can focus our efforts on making this part of our API super effective
+- R7: Disallow overrides in the MVP. All params are updated via config. This will change over time but I want users to tell which params they want to override (and how) instead of us making the assumption.
+
+
+## Design
+
+
+### Routes
+```yaml
+routes:
+    chat: /v1/language/{pool-id}/chat/ # or /v1/llms/...
+    transcribers: /v1/transcribers/{pool-id}/  # ...
+    speech-synthesizer:  /v1/synthesizers/{pool-id}/ # ...
+    multi-modal: /v1/multi/{pool-id}/multimodal/ # future route, not for MVP
+    embedding: /v1/embeddings/{pool-id}/embed/ # future route, not for MVP
+```
+
+
+#### User Request Schema for Chat Route
+
+Below is an example user request to the Chat route. The other routes will look very similar, if not identical. The provider specific params will look different.
+This request schema matches the OpenAI Chat request schema. The message field is required and the messageHistory field is optional. The role and content fields must be
+specified. These fields will be parsed when a specific provider is selected and messages will be passed in the correct schema to that provider.
+
+Unified API schema example:
+
+```json
+{
+  "message": [
+      {
+        "role": "user",
+        "content": "Where was it played?"
+      }
+    ],
+  "messageHistory": [
+    {"role": "system", "content": "You are a helpful assistant."},
+    {"role": "user", "content": "Who won the world series in 2020?"},
+    {"role": "assistant", "content": "The Los Angeles Dodgers won the World Series in 2020."},
+  ]
+}
+```
+If a user wants to specify different prompts for different models, they can do so.
+This assumes the following models have been added to the gateway via the config file.
+
+```json
+{
```

Comment:
@mkrueger12 regarding param overrides schema, I have a few points:
- Seems like this schema suggest to override all params for all models in a pool which may be tedious to do if you have several models in a pool and you need to override it for one of them.
- This schema is super different to the simple unified request schema. I would expect the unified request schema would allow to specify overrides (but it's just optional). This is not great because we need to have two schemas for one API instead of having it incorporated into the unified request schema
-It suggests to repass message history. Isn't this wasteful? Users most likely would have one message history despite models that were handling each specific request, wouldn't they?

Reply:
- Given Golang's validation patterns, we should probably follow the Kubernetes approach to allowing to pass different params. This is what I mean:

```
{
  "message": [
    {
      "role": "user",
      "content": "Where was it played?"
    }
  ],
  "messageHistory": [
    {"role": "system", "content": "You are a helpful assistant."},
    {"role": "user", "content": "Who won the world series in 2020?"},
    {"role": "assistant", "content": "The Los Angeles Dodgers won the World Series in 2020."},
  ],
  "overrides": {
    "openai": {
      "model-id1": {
        // params
      },
      "model-id2": {
        // params
      }
    },
    "cohere": {
      "model-id3": {
        // params
      }
    }
  }
}
```

Pay attention to that overrides field and it's structure. It contains provider keys that would help us to organize proper validation for each provider e.g. 

```go
type OpenAIOverride string {
	// proper validation of OpenAI specific params
}

type ParamOverrides struct {
	OpenAI *map[string]OpenAIOverride `yaml:"overrides"`
}

type RequestSchema struct {
	// ...
	Overrides *ParamOverrides `yaml:"overrides"`
}
```

We would simplify this a lot if there is away to unified incoming params completely. But something tells me there may be always some unique param users want to send to that specific provider.

Reply:
@roma-glushko I see what you're saying. The schema you've laid out seems reasonable but I want to be able to pass a different prompt if the model gets changed. Even serving the same prompt to models from the same provider is undesirable. 

When I think about overrides I am thinking more about the model hyper parameters than the prompt. We should, by default, allow the user to specify different prompts per model. How do you suggest we accomplish that?